File: Forge/src/main/java/hantonik/atomic/core/registration/FluidBuilder.java
package hantonik.atomic.core.registration;

import com.google.errorprone.annotations.CanIgnoreReturnValue;
import hantonik.atomic.core.client.model.fluid.ModelFluidType;
import lombok.RequiredArgsConstructor;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.Item;
import net.minecraft.world.level.block.LiquidBlock;
import net.minecraft.world.level.material.Fluid;
import net.minecraftforge.fluids.FluidType;
import net.minecraftforge.fluids.ForgeFlowingFluid;

import java.util.function.Supplier;

@RequiredArgsConstructor
public class FluidBuilder {
    private final FluidType.Properties properties;
    private final ResourceLocation location;

    private Supplier<? extends Item> bucket;
    private Supplier<? extends LiquidBlock> block;

    private boolean canConvertToSource = false;
    private float explosionResistance = 1.0F;

    private int slopeFindDistance = 4;
    private int levelDecreasePerBlock = 1;
    private int tickRate = 5;

    @CanIgnoreReturnValue
    public FluidBuilder canConvertToSource(boolean canConvertToSource) {
        this.canConvertToSource = canConvertToSource;

        return this;
    }

    @CanIgnoreReturnValue
    public FluidBuilder bucket(Supplier<? extends Item> bucket) {
        this.bucket = bucket;

        return this;
    }

    @CanIgnoreReturnValue
    public FluidBuilder block(Supplier<? extends LiquidBlock> block) {
        this.block = block;

        return this;
    }

    @CanIgnoreReturnValue
    public FluidBuilder slopeFindDistance(int slopeFindDistance) {
        this.slopeFindDistance = slopeFindDistance;

        return this;
    }

    @CanIgnoreReturnValue
    public FluidBuilder levelDecreasePerBlock(int levelDecreasePerBlock) {
        this.levelDecreasePerBlock = levelDecreasePerBlock;

        return this;
    }

    @CanIgnoreReturnValue
    public FluidBuilder explosionResistance(float explosionResistance) {
        this.explosionResistance = explosionResistance;

        return this;
    }

    @CanIgnoreReturnValue
    public FluidBuilder tickRate(int tickRate) {
        this.tickRate = tickRate;

        return this;
    }

    public ForgeFlowingFluid.Properties build(Supplier<? extends FluidType> type, Supplier<? extends Fluid> still, Supplier<? extends Fluid> flowing) {
        var properties = new ForgeFlowingFluid.Properties(type, still, flowing)
                .slopeFindDistance(this.slopeFindDistance)
                .levelDecreasePerBlock(this.levelDecreasePerBlock)
                .explosionResistance(this.explosionResistance)
                .tickRate(this.tickRate)
                .block(this.block)
                .bucket(this.bucket);

        if (this.canConvertToSource)
            this.properties.canConvertToSource(this.canConvertToSource);

        return properties;
    }

    public Supplier<? extends FluidType> getFluidType(Supplier<? extends Fluid> fluid, String descriptionId) {
        return () -> new ModelFluidType(this.properties.descriptionId(descriptionId), fluid);
    }

    public Supplier<? extends FluidType> getStillFluidType(Supplier<? extends Fluid> fluid) {
        return this.getFluidType(fluid, "fluid." + this.location.getNamespace() + "." + this.location.getPath());
    }

    public Supplier<? extends FluidType> getFlowingFluidType(Supplier<? extends Fluid> fluid) {
        return this.getFluidType(fluid, "fluid." + this.location.getNamespace() + ".flowing_" + this.location.getPath());
    }
}

File: Forge/src/main/kotlin/hantonik/atomic/core/registration/FluidBuilder.kt
package hantonik.atomic.core.registration

import com.google.errorprone.annotations.CanIgnoreReturnValue
import hantonik.atomic.core.client.model.fluid.ModelFluidType
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.item.Item
import net.minecraft.world.level.block.LiquidBlock
import net.minecraft.world.level.material.Fluid
import net.minecraftforge.fluids.FluidType
import net.minecraftforge.fluids.FluidType.Properties
import net.minecraftforge.fluids.ForgeFlowingFluid

open class FluidBuilder(private val properties: Properties, private val location: ResourceLocation) {
    private var bucket: (() -> Item)? = null
    private var block: (() -> LiquidBlock)? = null

    private var explosionResistance = 1.0F

    private var slopeFindDistance = 4
    private var levelDecreasePerBlock = 1
    private var tickRate = 5


    @CanIgnoreReturnValue
    fun bucket(bucket: () -> Item): FluidBuilder {
        this.bucket = bucket

        return this
    }

    @CanIgnoreReturnValue
    fun block(block: () -> LiquidBlock): FluidBuilder {
        this.block = block

        return this
    }

    @CanIgnoreReturnValue
    fun slopeFindDistance(slopeFindDistance: Int): FluidBuilder {
        this.slopeFindDistance = slopeFindDistance

        return this
    }

    @CanIgnoreReturnValue
    fun levelDecreasePerBlock(levelDecreasePerBlock: Int): FluidBuilder {
        this.levelDecreasePerBlock = levelDecreasePerBlock

        return this
    }

    @CanIgnoreReturnValue
    fun explosionResistance(explosionResistance: Float): FluidBuilder {
        this.explosionResistance = explosionResistance

        return this
    }

    @CanIgnoreReturnValue
    fun tickRate(tickRate: Int): FluidBuilder {
        this.tickRate = tickRate

        return this
    }

    open fun build(type: () -> FluidType, still: () -> Fluid, flowing: () -> Fluid): ForgeFlowingFluid.Properties {
        return ForgeFlowingFluid.Properties(type, still, flowing)
            .slopeFindDistance(this.slopeFindDistance)
            .levelDecreasePerBlock(this.levelDecreasePerBlock)
            .explosionResistance(this.explosionResistance)
            .tickRate(this.tickRate)
            .block(this.block)
            .bucket(this.bucket)
    }

    open fun getFluidType(fluid: () -> Fluid, descriptionId: String): () -> ModelFluidType = { ModelFluidType({ this.properties.descriptionId(descriptionId) }, fluid) }

    fun getStillFluidType(fluid: () -> Fluid): () -> ModelFluidType = this.getFluidType(fluid, "fluid.${this.location.namespace}.${this.location.path}")

    fun getFlowingFluidType(fluid: () -> Fluid): () -> ModelFluidType = this.getFluidType(fluid, "fluid.${this.location.namespace}.flowing_${this.location.path}")
}